doc/building.rst

```diff
@@ -15,88 +15,151 @@ These instructions are updated to work with version |version| of the Nextcloud C
 
 You have two possibilities to clone the repo.
 
-First option is As [remote URL](https://help.github.com/en/articles/which-remote-url-should-i-use) you can choose between cloning with HTTPS URL's, which is recommended or cloning with SSH URL's.
+First option is As RemoteURL_ you can choose between cloning with HTTPS URL's, which is recommended or cloning with SSH URL's.
 
-[https://github.com/nextcloud/desktop.git](https://github.com/nextcloud/desktop.git)
+.. _RemoteURL: https://help.github.com/en/articles/which-remote-url-should-i-use
+
+https://github.com/nextcloud/desktop.git
 
 When you don't have SSH key added to your GitHub account, than use HTTPS.
 
 When you no part of the nextcloud organisation, clone with HTTPS:
 
-```
-$ git clone git@github.com:nextcloud/desktop.git
-```
+.. code-block:: bash
+
+   $ git clone git@github.com:nextcloud/desktop.git
+
 
 macOS
 -----
 
-In addition to needing XCode (along with the command line tools), developing in
-the macOS environment requires extra dependencies.  You can install these
-dependencies through MacPorts_ or Homebrew_.  These dependencies are required
-only on the build machine, because non-standard libs are deployed in the app
-bundle.
+.. note:: While it is possible to do many of the following steps using GUI
+   frontends, wherever possible the Terminal commands are listed instead, in order
+   to streamline the process.
+
+1. Install Xcode from the Mac App Store:
+
+   https://apps.apple.com/app/xcode/id497799835
+
+Then, in Terminal:
+
+2. Install Xcode command line tools:
+
+   .. code-block:: bash
+
+      % xcode-select –install
+
+3. Install Homebrew from `brew.sh`_ (which will just give you the following):
+
+.. _`brew.sh`: https://brew.sh
+
+   .. code-block:: bash
+
+      % /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
+
+4. Fix Homebrew permissions. (This is `a known issue`_, and, yes, the workaround may lead to other shell warnings.)
+
+.. _`a known issue`: https://stackoverflow.com/a/63241724
+
+   .. code-block:: bash
+
+      % sudo chown -R $(whoami):admin /usr/local/* \
+         && sudo chmod -R g+rwx /usr/local/*
+
+5. Install Homebrew packages:
+
+   .. code-block:: bash
+
+      % brew install git qt qt5 qtkeychain cmake openssl pkgconfig glib doxygen cmocka
+
+6. Certain Homebrew packages are not automatically linked in places where
+   the build scripts can find them, so add them to your shell profile:
 
-The tested and preferred way to develop in this environment is through the use
-of HomeBrew_. The team has its own repository containing non-standard
-recipes.
+   .. code-block:: bash
 
-To set up your build environment for development using HomeBrew_:
+      % echo 'export OPENSSL_ROOT_DIR="/usr/local/opt/openssl@1.1"' >> ~/.zshrc
+      % echo 'export Qt5_DIR="/usr/local/opt/qt"' >> ~/.zshrc
+      % echo 'export Qt5LinguistTools_DIR="/usr/local/opt/qt/lib/cmake/Qt5LinguistTools/"' >> ~/.zshrc
```

Comment:
I would do
```
export OPENSSL_ROOT_DIR=$(brew --prefix openssl)
QT_INSTALL_DIR=$(brew --prefix qt)
export CMAKE_PREFIX_PATH=${QT_INSTALL_DIR}/lib/cmake/Qt5:${QT_INSTALL_DIR}/lib/cmake/Qt5LinguistTools:${CMAKE_PREFIX_PATH}
```
That way the CMake finds the correct Qt5 CMake config files.
I also would not put them into `.zshrc` you may want to develop other applications on your machine and don't want your environment polluted by one project. My suggestion would be to replace `~/.zshrc` with for example `~/prepare_desktop_dev_env.sh`. Every time you want to build you then need to source this file with `source ~/prepare_desktop_build_env.sh`. If this is to much work for you, you can use a tool like [direnv](https://direnv.net/) and place the `prepare_desktop_build_env.sh` in the desktop repo.

Reply:
I did this initially, but the problem with putting `$(brew --prefix <package>)` in your shell profile is that it runs every time you open a new shell, which can slow things down substantially. I don't know the differences between the various shell-profile files, though, so the loading time may be less of an issue if it can be made to be invoked less frequently.

FWIW you do know that I suggested just putting the calls to `brew --prefix` in the CMake file (which would reduce how often they run), but Kevin shot that down.

Reply:
> I did this initially, but the problem with putting $(brew --prefix <package>) in your shell profile is that it runs every time you open a new shell, which can slow things down substantially. I don't know the differences between the various shell-profile files, though, so the loading time may be less of an issue if it can be made to be invoked less frequently.

Yes that's one reason why I suggest that you don't put the statements into `.zshrc`. Then they will not get executed if not needed. If you use absolute paths you may run into problems if the path changes. With using brew prefix the instructions are more future proof. 

> FWIW you do know that I suggested just putting the calls to brew --prefix in the CMake file (which would reduce how often they run), but Kevin shot that down.

Yes, because that would make brew a requirement of an macos build but you don't need to use brew to build the client. You could also install the software by hand and maybe later we start to use a package manager like [Conan](https://conan.io/) to ship our build dependencies and then the CMake files would not work anymore. CMake is just a (meta) build system, not a package (dependency) manager. CMake's only task is it to create the correct build files. Managing dependencies is out of the scope of CMake :)

Reply:
My thought with these instructions is that it's more reasonable, for example, to explicitly tell people to use Homebrew than to bury that in the CMake. And the reason I put the paths in `.zshrc` is that it's tedious to have to reenter (i.e. copy and paste) the same information every time you run `cmake`. One option would be to tell the user (a) put it in a shell script, or (b) set it up in such-and-such a way on VS Code, but then that would assume the user is setting up VS Code or is otherwise going to set up a shell script.

How would you recommend I streamline the instructions so that the user doesn't have to run more than `cmake` and `make` every time they want to test an incrementally updated build (and assuming there could be a new shell session)? Again, the question is how you'd recommend **streamlining** the instructions.

Reply:
I should also clarify that I intend to modify the disclaimer at the top of the instructions and add a request that people open a GitHub Issue if the instructions don't work. This way, if the absolute path changes with future versions of Homebrew or whatever, the instructions can be updated to reflect that.

Reply:
Normally you run ` cmake` once when you clone the project. After that, you don't need to run it anymore unless you want to change the build configuration. With my proposal you would need to run `source ~/prepare_desktop_dev_env.sh` once a day if you open your terminal or if you open a new terminal in which you want to build. As I said if this is too much work for you, you can use a tool like direnv (I don't know if it's available for mac but anyway I would say that is out of the scope for the build instructions, that's a personal decision). 
Another idea: If your goal is to run the least possible commands, you can write your own shell script that exports the correct variables if they are not set and then runs cmake and make (that's what I do and it's really not difficult to write such a script. You could even start the app directly from that script. If you are really motivated, you could put such a script at the end of the build instructions;) ).

I would even say that it is harmful to put the instructions into `.zshrc` because you may get errors or side effects that you don't expect.

I think of the build instructions more like what steps need to be taken to build the app and not how your workflow should be because as you said people use different editors or IDEs. (If I were to write the build instructions I would not even put the export statements into a file, I would give the user a hint that she may want to put these instructions into a file to make her life easier) :)

Reply:
I just checked, and the following does in fact work:
```
% echo 'some string' >> ~/non_existent_file
```
Which is to say I could safely replace `~/.zshrc` with something like `~/.nextcloud_build_variables` (no extension needed, since it's not executable) and export them with `source ~/.nextcloud_build_variables` or `. ~/.nextcloud_build_variables` in a later step (i.e. right before running `cmake ..`). This would allow the environment to use up-to-the-minute `brew --prefix` paths while avoiding the necessity to run `brew --prefix` multiple times with every new shell session.

Of course, alternately, yes, the option for a build script could work, too.

> (If I were to write the build instructions I would not even put the export statements into a file, I would give the user a hint that she may want to put these instructions into a file to make her life easier)

I plan on having the build-instructions disclaimer explain that the instructions are not the only way of doing things; they are just a consistent and straightforward way of explaining how to do things.

Reply:
How does [this](https://github.com/nextcloud/desktop/pull/2931/commits/801d20e3b0c966132943b6a139df2871aedaf56d) look?

Reply:
it does look like a markdown file.

Reply:
> export them with `source ~/.nextcloud_build_variables` or `. ~/.nextcloud_build_variables` in a later step (i.e. right before running `cmake ..`). This would allow the environment to use up-to-the-minute `brew --prefix` paths while avoiding the necessity to run `brew --prefix` multiple times with every new shell session.

I think this text needs to be used in this section instead of just "The file ``~/.nextcloud_build_variables`` is just a suggestion for  convenience. You can use a different file or create an entire shell  script, but this way of doing things is the simplest to explain."